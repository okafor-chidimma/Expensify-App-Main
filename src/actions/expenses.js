import uuid from 'uuid';
// Add expense
export const addExpense = ({ amount = 0, createdAt = 0, note = '', description = '' } = {}) => (
  {
    type: 'ADD_EXPENSE',
    expense: {
      id: uuid(),
      amount,
      createdAt,
      description,
      note
    }
  }
);

// remove expense

export const removeExpense = ({ id }) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  }
};

// Edit Expense

export const editExpense = ({ id }, updatedExpense) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updatedExpense
  }
};
