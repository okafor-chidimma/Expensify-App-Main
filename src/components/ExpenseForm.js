import React, { Component } from 'react';
import 'react-dates/initialize';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends Component {
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: ''
  }
  handleAddDescription = (event) => {
    const description = event.target.value;
    this.setState(() => ({
      description
    }));
  };
  handleAddNote = (event) => {
    const note = event.target.value;
    this.setState(() => ({
      note
    }));
  };
  handleAddAmount = (event) => {
    const amount = event.target.value;
    // this condition is for users to be able to clear their input
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    } else {
      console.log('invalid amount');
      this.setState(() => ({ error: 'Invalid Input For Amount' }));
    }
  };
  onDateChange = (date) => {
    if (date) {
      this.setState({
        createdAt: date
      })
    }

  }
  onFocusChange = ({ focused }) => {
    this.setState({
      calendarFocused: focused
    })
  }
  onFormSubmit = (event) => {
    event.preventDefault();
    if (!this.state.description || !this.state.amount) {
      console.log('i should have error');
      this.setState(() => ({ error: 'Please neither amount nor description can be empty' }));
    } else {
      this.setState(() => ({ error: '' }));
      console.log('submitted');
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  }
  render() {
    return (
      <div>
        {
          this.state.error && <p style={{ color: 'red' }}>{this.state.error}</p>
        }
        <form onSubmit={this.onFormSubmit}>
          <input
            type="text"
            name="description"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.handleAddDescription}
          />
          <input
            type="text"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.handleAddAmount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Enter a note for the expense"
            value={this.state.note}
            onChange={this.handleAddNote}
          ></textarea>
          <button>
            {this.props.buttonState === "edit" ? 'Edit Expense' : 'Add Expense'}
          </button>
        </form>
      </div>
    )
  }
};

export default ExpenseForm;