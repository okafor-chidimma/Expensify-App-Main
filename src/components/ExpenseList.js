import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpense from '../selectors/expenses';

export const ExpenseList = ({ expenses }) => (
  <div>
    <h1>Expense List</h1>
    {
      expenses.length === 0 ? (
        <p>No expenses Yet</p>
      ) : (
          expenses.map((expense) => {
            return (
              <ExpenseListItem
                {...expense}
                key={expense.id}
              />
            )
          })
        )
    }
    {
      // props.expenses.map(({ description, amount, createdAt, id }) => {
      //   return (
      //     <ExpenseListItem
      //       description={description}
      //       amount={amount}
      //       createdAt={createdAt}
      //       key={id}
      //     />
      //   )
      // })

    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpense(state.expenses, state.filters)
  }
};

export default connect(mapStateToProps)(ExpenseList);


// it can re-written as

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses,
//     filters: state.filters

//   }
// })(ExpenseList);
// export default ConnectedExpenseList;
