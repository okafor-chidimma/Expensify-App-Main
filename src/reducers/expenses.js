// Expenses Reducer
const expenseStateDefault = [];
const expenseReducer = (state = expenseStateDefault, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updatedExpense
          }
        } else {
          return expense
        }
      });
    default:
      return state;
  }
};

export default expenseReducer;