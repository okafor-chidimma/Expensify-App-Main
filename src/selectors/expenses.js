import moment from 'moment';

// to sort the expense and return data that matches
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment) : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true;
    // this includes will always be true as long as the text is an empty string or a string contained in description
    // i.e textMatch will only be false is the text is not empty and cannot be found in description
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // console.log(textMatch, 'text match');
    // console.log(endDate, 'end date');
    // console.log(createdAtMoment, 'created at moment');
    // console.log(expense.createdAt, 'value of');
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1
    }
  });

}

export default getVisibleExpenses;