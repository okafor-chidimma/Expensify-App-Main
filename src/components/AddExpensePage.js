import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';
export class AddExpensePage extends React.Component {
  onSubmit = (expenseDetails) => {
    console.log(expenseDetails, 'i am expense details');
    this.props.addExpense(expenseDetails);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <h1>Add Expense Page</h1>
        <ExpenseForm
          buttonState="create"
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
}
const mapDispatchToProps = (dispatch) => (
  {
    addExpense: (expense) => dispatch(addExpense(expense))
  }
)
//             connect(mapStateToProps,mapDispatchToProps)
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
