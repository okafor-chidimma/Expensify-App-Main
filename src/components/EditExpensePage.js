import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  onSubmit = expenseDetails => {
    console.log(expenseDetails, 'i am expense details');
    // dispatch(editExpense({ id: props.match.params.id }, expense))
    this.props.startEditExpense(this.props.match.params.id, expenseDetails);
    this.props.history.push('/');
  };
  onButtonClick = () => {
    this.props.startRemoveExpense(this.props.match.params.id);
    // dispatch(removeExpense({ id: props.match.params.id }))
    this.props.history.push('/');
  };
  render() {
    // console.log(this.props);

    return (
      <div>
        Editing the expense of {this.props.match.params.id}
        <ExpenseForm
          expense={this.props.expense}
          buttonState='edit'
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onButtonClick}>Remove Item</button>
      </div>
    );
  }
}

// connect method has access to the connected component props by accepting a second parameter which is the props

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};
const mapDispatchToProps = dispatch => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: id => dispatch(startRemoveExpense(id))
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
