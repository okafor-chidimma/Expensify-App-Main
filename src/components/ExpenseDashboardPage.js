import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseSummary from "./ExpenseSummary";
const ExpenseDashBoardPage = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashBoardPage;
