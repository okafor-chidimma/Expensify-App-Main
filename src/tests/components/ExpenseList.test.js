import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

describe('Rendering the Expense List component', () => {
  it('Render the <ExpenseList/> with expenses as props', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('Render <ExpenseList/> with an empty expenses array as props', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  })
})