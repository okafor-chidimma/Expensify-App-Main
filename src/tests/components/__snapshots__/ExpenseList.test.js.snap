// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering the Expense List component Render <ExpenseList/> with an empty expenses array as props 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No expenses Yet
  </p>
</div>
`;

exports[`Rendering the Expense List component Render the <ExpenseList/> with expenses as props 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={54500}
    createdAt={0}
    description="January rent"
    id="1"
    key="1"
    note="Last rent before moving"
  />
  <ExpenseListItem
    amount={4500}
    createdAt={-345600000}
    description="Cable Money"
    id="2"
    key="2"
    note="Gotv cable money"
  />
  <ExpenseListItem
    amount={45000}
    createdAt={345600000}
    description="Data Money"
    id="3"
    key="3"
    note="Subscription for the month"
  />
</div>
`;
